detector/calorimeter/HCalThreePartsEndcap_o1_v01_geo.cpp

```diff
@@ -1,467 +1,496 @@
-  std::vector<std::vector<dd4hep::PlacedVolume> > seqInLayers;
-  seqInLayers.reserve(layerDepths1.size()+layerDepths2.size()+layerDepths3.size());
-  std::vector<dd4hep::PlacedVolume> tilesPerLayer;
-  tilesPerLayer.reserve(layerDepths1.size()+layerDepths2.size()+layerDepths3.size());
-
-  // loop over R ("layers")
-  double layerR = 0.;
-  for (unsigned int idxLayer = 0; idxLayer < layerDepths1.size(); ++idxLayer) {
-    // in Module rmin = 0  for first wedge, changed radius to the full radius starting at (0,0,0)
-    double rminLayer = sensitiveBarrel1Rmin + layerR;
-    double rmaxLayer = sensitiveBarrel1Rmin + layerR + layerDepths1.at(idxLayer);
-    layerR += layerDepths1.at(idxLayer);
-
-    //alternate: even layers consist of tile sequence b, odd layer of tile sequence a
-    unsigned int sequenceIdx = (idxLayer+1) % 2;
-    
-    dd4hep::Tube tileSequenceShape(rminLayer, rmaxLayer, 0.5*dzSequence);
-    Volume tileSequenceVolume("HCalECTileSequenceVol1", tileSequenceShape, aLcdd.air());
-
-    lLog << MSG::DEBUG << "layer radii:  " << rminLayer << " - " << rmaxLayer << " [cm]" << endmsg;
-    
-
-    dd4hep::Tube layerShape(rminLayer, rmaxLayer, dzDetector1 );
-    Volume layerVolume("HCalECLayerVol1", layerShape, aLcdd.air());
-
   
-    layerVolume.setVisAttributes(aLcdd.invisible());
-    unsigned int idxSubMod = 0;
-
-    dd4hep::Position moduleOffset1 (0,0,sign * extBarrelOffset1);
-
-    dd4hep::PlacedVolume placedLayerVolume = aEnvelope.placeVolume(layerVolume, moduleOffset1);
-    unsigned int type1 = 0;
-    if (sign<0) {
-      type1 = 3;
-    }
-    placedLayerVolume.addPhysVolID("type", type1);  // First module type=0,3 in front of second +/-
-    placedLayerVolume.addPhysVolID("layer", idxLayer);
-    layers.push_back(placedLayerVolume);
-   
-    double tileZOffset = - 0.5* dzSequence;
-
-    // first Z loop (tiles that make up a sequence)
-    for (xml_coll_t xCompColl(sequences[sequenceIdx], _Unicode(module_component)); xCompColl;
-	 ++xCompColl, ++idxSubMod) {
-      xml_comp_t xComp = xCompColl;
-      dd4hep::Tube tileShape(rminLayer, rmaxLayer, 0.5 * xComp.thickness());
-      
-      Volume tileVol("HCalECTileVol_"+ xComp.materialStr()
-		     , tileShape,
-		     aLcdd.material(xComp.materialStr()));
-      tileVol.setVisAttributes(aLcdd, xComp.visStr());
-      
-      dd4hep::Position tileOffset(0, 0, tileZOffset + 0.5 * xComp.thickness() );
-      dd4hep::PlacedVolume placedTileVol = tileSequenceVolume.placeVolume(tileVol, tileOffset);
-      
-      if (xComp.isSensitive()) {
-        tileVol.setSensitiveDetector(sensDet);
-        tilesPerLayer.push_back(placedTileVol);
-      }
-      tileZOffset += xComp.thickness();
+    for (std::vector<xml_comp_t>::iterator it = Layers.begin(); it != Layers.end(); ++it){ 
+        xml_comp_t layer = *it;
+        Dimension layerDimension(layer.dimensions());
+        numSequencesR1 += layerDimension.nmodules();
+        numSequencesR2 += layerDimension.nModules();
```

Comment:
I must say that having two functions (and two xml tags) whose name differs only by the capitalisation of a letter is not a very good idea... have you considered having more expressive names?

Reply:
Agreed, I replaced nModules by nsegments as in the current implementation, the name needs to be a member of struct dd4hep::xml::Dimension. So it's not really easy to find something more meaningful.

Reply:
> the name needs to be a member of struct dd4hep::xml::Dimension

Hmm,

```
layerDimension.child(_U(whatever))
```

should work?

Reply:
thanks @andresailer , I'd prefer to keep it is as it for now and will try to implement this in the next version of the code

Reply:
Apologies for giving incorrect suggestions!
```c++
    const double whatever    = dimensions.attr< int >("whatever");
```
This actually compiles